### PR TITLE
[DataGrid] Fix cell value type in quick filtering v7

### PR DESCRIPTION
**Pull request by cherniavskii**
Reverts https://github.com/mui/mui-x/pull/10569/commits/c14f23c461fbf6782ed35bd3e752c1c3e2853566 from https://github.com/mui/mui-x/pull/10569
I thought this was an optimization, but it turned out to cause issues with non-string column types.
E.g. for the `date` column type, `value` is supposed to be of type `Date`, while it currently is `string` (because value formatter returns string).

I have discovered this issue while working on https://github.com/mui/mui-x/issues/10581

### CODE REVIEW

**Review by MBilalShafi — APPROVED** [review]

File: packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts [hunk 1]
```diff
@@ -220,7 +221,7 @@ const getFilterCallbackFromItem = (
         if (ignoreDiacritics) {
           value = removeDiacritics(value);
         }
-        return applyFilterOnRow(value, row, column, apiRef);
+        return applyFilterOnRow(value, row, column, getPublicApiRef(apiRef));
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts:224]: We were passing private apiRef here

File: packages/grid/x-data-grid/src/colDef/gridStringOperators.ts [hunk 2]
```diff
@@ -5,14 +5,19 @@ import { GridFilterItem } from '../models/gridFilterItem';
...
     if (!value) {
       return null;
     }
     const filterRegex = new RegExp(escapeRegExp(value), 'i');
-    return (columnValue): boolean => {
+    return (_, row, column, apiRef): boolean => {
+      let columnValue = apiRef.current.getRowFormattedValue(row, column);
+      if (apiRef.current.ignoreDiacritics) {
+        columnValue = removeDiacritics(columnValue);
+      }
```
Comment (cherniavskii) [inline packages/grid/x-data-grid/src/colDef/gridStringOperators.ts:20]: This is a disadvantage compared to v6 filtering, where it was possible to overwrite the `cellParams.formattedValue`.

File: packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts (hunk 1 above)
Comment (MBilalShafi) [reply packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts:224]: Oops, nice catch 👍

Comment (romgrk) [reply packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts:224]: Can we create the API ref outside this function? I think this one is inside the filter loop, so we're doing `O(n)` allocations if we call `getPublicApiRef` here.

Comment (romgrk) [reply packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts:224]: Same comment applies below, line 420.

**Review by romgrk — APPROVED** [review]

File: packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts (hunk 1 above)
Comment (cherniavskii) [reply packages/grid/x-data-grid/src/hooks/features/filter/gridFilterUtils.ts]: Thanks for catching it, updated it!

**Pull request merged**